finagle-netty4/src/main/scala/com/twitter/finagle/netty4/ssl/client/SslClientVerificationHandler.scala

```diff
@@ -41,20 +40,21 @@ private[netty4] class SslClientVerificationHandler(
   }
 
   private[this] def verifySession(session: SSLSession, ctx: ChannelHandlerContext): Unit = {
```

Comment:
I think this needs to be async, and we need to wait on removing the handler until we have a result from the verification.  should we add a test for this?

Reply:
+1 pending writes will get written while the client is still trying to determine whether the remote peer is authorized.

Reply:
Is the following correct?

```scala
private[this] def verifySession(session: SSLSession, ctx: ChannelHandlerContext): Future[Boolean] = {
  sessionVerifier(address, config, session).respond {
    case Return(verifierResult) =>
      if (!verifierResult) {
        fail(
          SslVerificationFailedException(
            Some(new Exception("Failed client verification")),
            inet
          )
        )
        ctx.close()
      } // if verification is successful, do nothing here
    case Throw(e) =>
      fail(SslVerificationFailedException(Some(e), inet))
  }
}

override def handlerAdded(ctx: ChannelHandlerContext): Unit = {
  // React on satisfied handshake promise.
  sslHandler
    .handshakeFuture()
    .addListener(new GenericFutureListener[NettyFuture[Channel]] {
      def operationComplete(f: NettyFuture[Channel]): Unit =
        if (f.isSuccess) {
          if (sslHandler.engine.isInboundDone) {
            // Likely that the server failed to verify the client or the handshake failed in an unexpected way.
            fail(
              new SslVerificationFailedException(
                Some(new Exception("Failed server verification")),
                inet
              )
            )
          } else {
            val session = sslHandler.engine().getSession
            verifySession(session, ctx).respond { _ =>
              if (onHandshakeComplete.setDone()) {
                // If the original promise has not yet been satisfied, it can now be marked as successful,
                // and the `SslClientVerificationHandler` can be removed from the pipeline. If the promise is
                // already satisfied, the connection was failed and the failure has already been
                // propagated to the dispatcher so we don't need to worry about cleaning up the pipeline.
                ctx.pipeline.remove(self) // drains pending writes when removed
              }
            }
          }
        } else {
          // Handshake promise is failed so given `SslHandler` is going to close the channel we only
          // need to fail pending writes and the connect promise.
          fail(f.cause())
        }
    })

  super.handlerAdded(ctx)
}
```

Reply:
The `respond` within the `verifySession` method looks correct, but the one outside of it is not. There it's fine to use `onSuccess` in this instance:

    ...
    verifySession(session, ctx).onSuccess { _ => 
        if (on...setDone()) {
          ...
          ctx.pipeline.remove(self) ...
       }
    }
    ...